electrumsv/network_support/esv_client.py

```diff
@@ -0,0 +1,436 @@
+            yield tsc_merkle_proof
+
+    async def subscribe_to_general_notifications(self) -> AsyncIterable[GeneralNotification]:
+        """Concurrent fetching of peer channel messages is left to the caller in order to keep
+        this class very simple"""
+        ws_base_url = self._replace_http_with_ws(self.base_url)
+        url = ws_base_url + "api/v1/web-socket" + f"?token={self.master_token}"
+        logger.debug(f"URL IS: {url}")
+        try:
+            async with self.session.ws_connect(url, headers={}, timeout=5.0) as ws:
+                logger.info(f'Connected to {url}')
+                msg: aiohttp.WSMessage
+                async for msg in ws:
+                    if msg.type == aiohttp.WSMsgType.TEXT:
+                        notification: GeneralNotification = json.loads(msg.data)
+                        yield notification
+
+                    if msg.type in (aiohttp.WSMsgType.CLOSE, aiohttp.WSMsgType.ERROR,
+                            aiohttp.WSMsgType.CLOSED, aiohttp.WSMsgType.CLOSING):
+                        logger.info("CLOSED")
+                        break
+        except WSServerHandshakeError as e:
+            if e.status == 401:
+                raise WebsocketUnauthorizedException()
+
+    # ----- HeaderSV APIs ----- #
+    async def get_single_header(self, block_hash: bytes) -> bytes:
+        url = self.base_url + f"api/v1/headers/{hash_to_hex_str(block_hash)}"
+        headers = {}
+        headers.update(self.headers)
+        headers.update({"Accept": "application/octet-stream"})
+        async with self.session.get(url, headers=headers) as resp:
+            resp.raise_for_status()
+            raw_header = await resp.read()
+            return raw_header
+
+    async def get_headers_by_height(self, from_height: int, count: Optional[int]=None) \
+            -> bytes:
+        url = self.base_url + "api/v1/headers/by-height" + f"?height={from_height}"
+        if count:
+            url += f"&count={count}"
+        headers = {}
+        headers.update(self.headers)
+        headers.update({"Accept": "application/octet-stream"})
+        async with self.session.get(url, headers=headers) as resp:
+            resp.raise_for_status()
+            raw_headers_array = await resp.read()
+            return raw_headers_array
+
+    async def get_chain_tips(self) -> TipResponse:
+        url = self.base_url + "api/v1/headers/tips"
+        headers = {}
+        headers.update(self.headers)
+        headers.update({"Accept": "application/json"})
+        async with self.session.get(url, headers=self.headers) as resp:
+            resp.raise_for_status()
+            json_tip_response: TipResponse = await resp.json()
+            return json_tip_response
+
+    async def subscribe_to_headers(self) -> AsyncIterable[TipResponse]:
+        ws_base_url = self._replace_http_with_ws(self.base_url)
+        url = ws_base_url + "api/v1/headers/tips/websocket"
+
+        logger.debug(f"URL IS: {url}")
```

Comment:
Better log message, lots of these log messages have no context to indicate why they are being logged.